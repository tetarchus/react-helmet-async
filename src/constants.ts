const TAG_PROPERTIES = {
  CHARSET: 'charset',
  CSS_TEXT: 'cssText',
  HREF: 'href',
  HTTPEQUIV: 'http-equiv',
  INNER_HTML: 'innerHTML',
  ITEM_PROP: 'itemprop',
  NAME: 'name',
  PROPERTY: 'property',
  REL: 'rel',
  SRC: 'src',
};

const ATTRIBUTE_NAMES = {
  BODY: 'bodyAttributes',
  HTML: 'htmlAttributes',
  TITLE: 'titleAttributes',
};

const TAG_NAMES = {
  BASE: 'base',
  BODY: 'body',
  HEAD: 'head',
  HTML: 'html',
  LINK: 'link',
  META: 'meta',
  NOSCRIPT: 'noscript',
  SCRIPT: 'script',
  STYLE: 'style',
  TITLE: 'title',
  FRAGMENT: 'Symbol(react.fragment)',
};

const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
};

const VALID_TAG_NAMES = Object.keys(TAG_NAMES).map((tagName) => TAG_NAMES[tagName] as string);

const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
};

const HTML_TAG_MAP = Object.fromEntries(
  Object.keys(REACT_TAG_MAP).map((key) => [REACT_TAG_MAP[key as keyof typeof REACT_TAG_MAP], key]),
);

const HELMET_ATTRIBUTE = 'data-rh';

const HELMET_PROPS = {
  DEFAULT_TITLE: 'defaultTitle',
  DEFER: 'defer',
  ENCODE_SPECIAL_CHARACTERS: 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE: 'onChangeClientState',
  TITLE_TEMPLATE: 'titleTemplate',
  PRIORITIZE_SEO_TAGS: 'prioritizeSeoTags',
};

export {
  ATTRIBUTE_NAMES,
  HELMET_ATTRIBUTE,
  HELMET_PROPS,
  HTML_TAG_MAP,
  REACT_TAG_MAP,
  SEO_PRIORITY_TAGS,
  TAG_NAMES,
  TAG_PROPERTIES,
  VALID_TAG_NAMES,
};
