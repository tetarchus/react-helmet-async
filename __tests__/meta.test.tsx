import React from 'react';

import { Helmet } from '..';
import { HELMET_ATTRIBUTE } from '../components/constants';

import { ReactServer } from './setup/encoderFix';
import { after, before, isArray, render } from './setup/testSetup';

describe('<meta> tags', () => {
  beforeEach(before);
  afterEach(after);

  describe('Browser', () => {
    describe('API', () => {
      it('updates meta tags', () => {
        expect.assertions(2);

        render(
          <Helmet
            defer={false}
            meta={[
              { charset: 'utf-8' },
              {
                name: 'description',
                content: 'Test description',
              },
              {
                'http-equiv': 'content-type',
                content: 'text/html',
              },
              { property: 'og:type', content: 'article' },
              { itemprop: 'name', content: 'Test name itemprop' },
            ]}
          />,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];

        expect(existingTags).toBeDefined();

        const filteredTags = existingTags.filter(
          (tag) =>
            tag.getAttribute('charset') === 'utf-8' ||
            (tag.getAttribute('name') === 'description' &&
              tag.getAttribute('content') === 'Test description') ||
            (tag.getAttribute('http-equiv') === 'content-type' &&
              tag.getAttribute('content') === 'text/html') ||
            (tag.getAttribute('itemprop') === 'name' &&
              tag.getAttribute('content') === 'Test name itemprop'),
        );

        expect(filteredTags.length).toBeGreaterThanOrEqual(4);
      });

      it('clears all meta tags if none are specified', () => {
        expect.assertions(2);

        render(
          <Helmet defer={false} meta={[{ name: 'description', content: 'Test description' }]} />,
        );

        render(<Helmet defer={false} />);

        const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(0);
      });

      it('does not accept tags without "name", "http-equiv", "property", "charset", or "itemprop"', () => {
        expect.assertions(2);
        // Testing invalid properties
        // @ts-expect-error - 'href' does not exist in type...
        render(<Helmet defer={false} meta={[{ href: "won't work" }]} />);

        const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(0);
      });

      it('sets meta tags based on deepest nested component', () => {
        expect.assertions(16);

        render(
          <div>
            <Helmet
              defer={false}
              meta={[
                { charset: 'utf-8' },
                {
                  name: 'description',
                  content: 'Test description',
                },
              ]}
            />
            <Helmet
              defer={false}
              meta={[
                {
                  name: 'description',
                  content: 'Inner description',
                },
                { name: 'keywords', content: 'test,meta,tags' },
              ]}
            />
          </div>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag, secondTag, thirdTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(3);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('charset')).toBe('utf-8');
        expect(firstTag.outerHTML).toMatchSnapshot();

        expect(secondTag).toBeInstanceOf(Element);
        expect(secondTag.getAttribute).toBeDefined();
        expect(secondTag.getAttribute('name')).toBe('description');
        expect(secondTag.getAttribute('content')).toBe('Inner description');
        expect(secondTag.outerHTML).toMatchSnapshot();

        expect(thirdTag).toBeInstanceOf(Element);
        expect(thirdTag.getAttribute).toBeDefined();
        expect(thirdTag.getAttribute('name')).toBe('keywords');
        expect(thirdTag.getAttribute('content')).toBe('test,meta,tags');
        expect(thirdTag.outerHTML).toMatchSnapshot();
      });

      it('allows duplicate meta tags if specified in the same component', () => {
        expect.assertions(12);

        render(
          <Helmet
            defer={false}
            meta={[
              {
                name: 'description',
                content: 'Test description',
              },
              {
                name: 'description',
                content: 'Duplicate description',
              },
            ]}
          />,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag, secondTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(2);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('name')).toBe('description');
        expect(firstTag.getAttribute('content')).toBe('Test description');
        expect(firstTag.outerHTML).toMatchSnapshot();

        expect(secondTag).toBeInstanceOf(Element);
        expect(secondTag.getAttribute).toBeDefined();
        expect(secondTag.getAttribute('name')).toBe('description');
        expect(secondTag.getAttribute('content')).toBe('Duplicate description');
        expect(secondTag.outerHTML).toMatchSnapshot();
      });

      it('overrides duplicate meta tags with single meta tag in a nested component', () => {
        expect.assertions(7);

        render(
          <div>
            <Helmet
              defer={false}
              meta={[
                {
                  name: 'description',
                  content: 'Test description',
                },
                {
                  name: 'description',
                  content: 'Duplicate description',
                },
              ]}
            />
            <Helmet
              defer={false}
              meta={[
                {
                  name: 'description',
                  content: 'Inner description',
                },
              ]}
            />
          </div>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(1);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('name')).toBe('description');
        expect(firstTag.getAttribute('content')).toBe('Inner description');
        expect(firstTag.outerHTML).toMatchSnapshot();
      });

      it('overrides single meta tag with duplicate meta tags in a nested component', () => {
        expect.assertions(12);

        render(
          <div>
            <Helmet
              defer={false}
              meta={[
                {
                  name: 'description',
                  content: 'Test description',
                },
              ]}
            />
            <Helmet
              defer={false}
              meta={[
                {
                  name: 'description',
                  content: 'Inner description',
                },
                {
                  name: 'description',
                  content: 'Inner duplicate description',
                },
              ]}
            />
          </div>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag, secondTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(2);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('name')).toBe('description');
        expect(firstTag.getAttribute('content')).toBe('Inner description');
        expect(firstTag.outerHTML).toMatchSnapshot();

        expect(secondTag).toBeInstanceOf(Element);
        expect(secondTag.getAttribute).toBeDefined();
        expect(secondTag.getAttribute('name')).toBe('description');
        expect(secondTag.getAttribute('content')).toBe('Inner duplicate description');
        expect(secondTag.outerHTML).toMatchSnapshot();
      });

      it('does not render tag when primary attribute is null', () => {
        expect.assertions(1);

        render(
          <Helmet
            defer={false}
            meta={[
              // Testing invalid values
              // @ts-expect-error - Type 'undefined' is not assignable to type 'string'
              {
                name: undefined,
                content: 'Inner duplicate description',
              },
            ]}
          />,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes);

        expect(existingTags).toHaveLength(0);
      });

      it('fails gracefully when meta is wrong shape', () => {
        expect.assertions(3);

        const originalConsole = global.console;
        jest.spyOn(global.console, 'warn').mockImplementation();
        jest.spyOn(global.console, 'error').mockImplementation();
        // Testing invalid type
        // @ts-expect-error - 'name' does not exist in type
        render(<Helmet defer={false} meta={{ name: 'title', content: 'some title' }} />);

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes);

        expect(existingTags).toHaveLength(0);

        expect(global.console.warn).toHaveBeenCalledWith(
          'Helmet: meta should be of type "Array". Instead found type "object"',
        );

        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        expect((global.console.warn as jest.Mock).mock.calls[0][0]).toMatchSnapshot();

        global.console = originalConsole;
      });
    });

    describe('Declarative API', () => {
      it('updates meta tags', () => {
        expect.assertions(2);

        render(
          <Helmet defer={false}>
            <meta charSet='utf-8' />
            <meta name='description' content='Test description' />
            <meta httpEquiv='content-type' content='text/html' />
            <meta property='og:type' content='article' />
            <meta itemProp='name' content='Test name itemprop' />
          </Helmet>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];

        expect(existingTags).toBeDefined();

        const filteredTags = existingTags.filter(
          (tag) =>
            tag.getAttribute('charset') === 'utf-8' ||
            (tag.getAttribute('name') === 'description' &&
              tag.getAttribute('content') === 'Test description') ||
            (tag.getAttribute('http-equiv') === 'content-type' &&
              tag.getAttribute('content') === 'text/html') ||
            (tag.getAttribute('itemprop') === 'name' &&
              tag.getAttribute('content') === 'Test name itemprop'),
        );

        expect(filteredTags.length).toBeGreaterThanOrEqual(4);
      });

      it('clears all meta tags if none are specified', () => {
        expect.assertions(2);

        render(
          <Helmet defer={false}>
            <meta name='description' content='Test description' />
          </Helmet>,
        );

        render(<Helmet defer={false} />);

        const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(0);
      });

      it('does not accept tags without "name", "http-equiv", "property", "charset", or "itemprop"', () => {
        expect.assertions(2);

        render(
          <Helmet defer={false}>
            {/* Testing invalid properties
          @ts-expect-error - Property 'href' does not exist on type ... */}
            <meta href="won't work" />
          </Helmet>,
        );

        const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(0);
      });

      it('sets meta tags based on deepest nested component', () => {
        expect.assertions(16);

        render(
          <div>
            <Helmet defer={false}>
              <meta charSet='utf-8' />
              <meta name='description' content='Test description' />
            </Helmet>
            <Helmet defer={false}>
              <meta name='description' content='Inner description' />
              <meta name='keywords' content='test,meta,tags' />
            </Helmet>
          </div>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag, secondTag, thirdTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(3);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('charset')).toBe('utf-8');
        expect(firstTag.outerHTML).toMatchSnapshot();

        expect(secondTag).toBeInstanceOf(Element);
        expect(secondTag.getAttribute).toBeDefined();
        expect(secondTag.getAttribute('name')).toBe('description');
        expect(secondTag.getAttribute('content')).toBe('Inner description');
        expect(secondTag.outerHTML).toMatchSnapshot();

        expect(thirdTag).toBeInstanceOf(Element);
        expect(thirdTag.getAttribute).toBeDefined();
        expect(thirdTag.getAttribute('name')).toBe('keywords');
        expect(thirdTag.getAttribute('content')).toBe('test,meta,tags');
        expect(thirdTag.outerHTML).toMatchSnapshot();
      });

      it('allows duplicate meta tags if specified in the same component', () => {
        expect.assertions(12);

        render(
          <Helmet defer={false}>
            <meta name='description' content='Test description' />
            <meta name='description' content='Duplicate description' />
          </Helmet>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag, secondTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(2);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('name')).toBe('description');
        expect(firstTag.getAttribute('content')).toBe('Test description');
        expect(firstTag.outerHTML).toMatchSnapshot();

        expect(secondTag).toBeInstanceOf(Element);
        expect(secondTag.getAttribute).toBeDefined();
        expect(secondTag.getAttribute('name')).toBe('description');
        expect(secondTag.getAttribute('content')).toBe('Duplicate description');
        expect(secondTag.outerHTML).toMatchSnapshot();
      });

      it('overrides duplicate meta tags with single meta tag in a nested component', () => {
        expect.assertions(7);

        render(
          <div>
            <Helmet defer={false}>
              <meta name='description' content='Test description' />
              <meta name='description' content='Duplicate description' />
            </Helmet>
            <Helmet defer={false}>
              <meta name='description' content='Inner description' />
            </Helmet>
          </div>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(1);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('name')).toBe('description');
        expect(firstTag.getAttribute('content')).toBe('Inner description');
        expect(firstTag.outerHTML).toMatchSnapshot();
      });

      it('overrides single meta tag with duplicate meta tags in a nested component', () => {
        expect.assertions(12);

        render(
          <div>
            <Helmet defer={false}>
              <meta name='description' content='Test description' />
            </Helmet>
            <Helmet defer={false}>
              <meta name='description' content='Inner description' />
              <meta name='description' content='Inner duplicate description' />
            </Helmet>
          </div>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes) as Element[];
        const [firstTag, secondTag] = existingTags;

        expect(existingTags).toBeDefined();
        expect(existingTags).toHaveLength(2);

        expect(firstTag).toBeInstanceOf(Element);
        expect(firstTag.getAttribute).toBeDefined();
        expect(firstTag.getAttribute('name')).toBe('description');
        expect(firstTag.getAttribute('content')).toBe('Inner description');
        expect(firstTag.outerHTML).toMatchSnapshot();

        expect(secondTag).toBeInstanceOf(Element);
        expect(secondTag.getAttribute).toBeDefined();
        expect(secondTag.getAttribute('name')).toBe('description');
        expect(secondTag.getAttribute('content')).toBe('Inner duplicate description');
        expect(secondTag.outerHTML).toMatchSnapshot();
      });

      it('does not render tag when primary attribute is null', () => {
        expect.assertions(1);

        render(
          <Helmet defer={false}>
            <meta name={undefined} content='Inner duplicate description' />
          </Helmet>,
        );

        const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
        const existingTags = Array.prototype.slice.call(tagNodes);

        expect(existingTags).toHaveLength(0);
      });
    });
  });

  describe('Server', () => {
    describe('API', () => {
      it('renders meta tags as React components', () => {
        expect.assertions(10);

        const { context } = render(
          <Helmet
            defer={false}
            meta={[
              { charset: 'utf-8' },
              {
                name: 'description',
                content: 'Test description & encoding of special characters like \' " > < `',
              },
              { 'http-equiv': 'content-type', content: 'text/html' },
              { property: 'og:type', content: 'article' },
              { itemprop: 'name', content: 'Test name itemprop' },
            ]}
          />,
          true,
        );

        const head = 'helmet' in context ? context.helmet : null;

        expect(head?.meta).toBeDefined();
        expect(head?.meta.toComponent).toBeDefined();

        const metaComponent = head?.meta.toComponent() ?? [];

        expect(metaComponent).toStrictEqual(isArray);
        expect(metaComponent).toHaveLength(5);

        for (const meta of metaComponent) {
          expect(meta).toStrictEqual(expect.objectContaining({ type: 'meta' }));
        }

        const view = ReactServer.renderToStaticMarkup(<>{metaComponent}</>);

        expect(view).toMatchSnapshot();
      });

      it('renders meta tags as string', () => {
        expect.assertions(3);

        const { context } = render(
          <Helmet
            defer={false}
            meta={[
              { charset: 'utf-8' },
              {
                name: 'description',
                content: 'Test description & encoding of special characters like \' " > < `',
              },
              { 'http-equiv': 'content-type', content: 'text/html' },
              { property: 'og:type', content: 'article' },
              { itemprop: 'name', content: 'Test name itemprop' },
            ]}
          />,
          true,
        );

        const head = 'helmet' in context ? context.helmet : null;

        expect(head?.meta).toBeDefined();
        expect(head?.meta.toString).toBeDefined();
        expect(head?.meta.toString()).toMatchSnapshot();
      });
    });

    describe('Declarative API', () => {
      it('renders meta tags as React components', () => {
        expect.assertions(10);

        const { context } = render(
          <Helmet defer={false}>
            <meta charSet='utf-8' />
            <meta
              name='description'
              content={'Test description & encoding of special characters like \' " > < `'}
            />
            <meta httpEquiv='content-type' content='text/html' />
            <meta property='og:type' content='article' />
            <meta itemProp='name' content='Test name itemprop' />
          </Helmet>,
          true,
        );

        const head = 'helmet' in context ? context.helmet : null;

        expect(head?.meta).toBeDefined();
        expect(head?.meta.toComponent).toBeDefined();

        const metaComponent = head?.meta.toComponent() ?? [];

        expect(metaComponent).toStrictEqual(isArray);
        expect(metaComponent).toHaveLength(5);

        for (const meta of metaComponent) {
          expect(meta).toStrictEqual(expect.objectContaining({ type: 'meta' }));
        }

        const view = ReactServer.renderToStaticMarkup(<>{metaComponent}</>);

        expect(view).toMatchSnapshot();
      });

      it('renders meta tags as string', () => {
        expect.assertions(3);

        const { context } = render(
          <Helmet defer={false}>
            <meta charSet='utf-8' />
            <meta
              name='description'
              content='Test description &amp; encoding of special characters like &#x27; " &gt; &lt; `'
            />
            <meta httpEquiv='content-type' content='text/html' />
            <meta property='og:type' content='article' />
            <meta itemProp='name' content='Test name itemprop' />
          </Helmet>,
          true,
        );

        const head = 'helmet' in context ? context.helmet : null;

        expect(head?.meta).toBeDefined();
        expect(head?.meta.toString).toBeDefined();
        expect(head?.meta.toString()).toMatchSnapshot();
      });
    });
  });
});
