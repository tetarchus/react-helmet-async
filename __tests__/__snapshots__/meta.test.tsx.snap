// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<meta> tags Browser API allows duplicate meta tags if specified in the same component 1`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser API allows duplicate meta tags if specified in the same component 2`] = `"<meta name=\\"description\\" content=\\"Duplicate description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser API fails gracefully when meta is wrong shape 1`] = `"Helmet: meta should be of type \\"Array\\". Instead found type \\"object\\""`;

exports[`<meta> tags Browser API overrides duplicate meta tags with single meta tag in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser API overrides single meta tag with duplicate meta tags in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser API overrides single meta tag with duplicate meta tags in a nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner duplicate description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser API sets meta tags based on deepest nested component 1`] = `"<meta charset=\\"utf-8\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser API sets meta tags based on deepest nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser API sets meta tags based on deepest nested component 3`] = `"<meta name=\\"keywords\\" content=\\"test,meta,tags\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API allows duplicate meta tags if specified in the same component 1`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API allows duplicate meta tags if specified in the same component 2`] = `"<meta name=\\"description\\" content=\\"Duplicate description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API overrides duplicate meta tags with single meta tag in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API overrides single meta tag with duplicate meta tags in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API overrides single meta tag with duplicate meta tags in a nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner duplicate description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API sets meta tags based on deepest nested component 1`] = `"<meta charset=\\"utf-8\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API sets meta tags based on deepest nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Browser Declarative API sets meta tags based on deepest nested component 3`] = `"<meta name=\\"keywords\\" content=\\"test,meta,tags\\" data-rh=\\"true\\">"`;

exports[`<meta> tags Server API renders meta tags as React components 1`] = `"<meta data-rh=\\"true\\" itemProp=\\"name\\" content=\\"Test name itemprop\\"/><meta data-rh=\\"true\\" property=\\"og:type\\" content=\\"article\\"/><meta data-rh=\\"true\\" http-equiv=\\"content-type\\" content=\\"text/html\\"/><meta data-rh=\\"true\\" name=\\"description\\" content=\\"Test description &amp; encoding of special characters like &#x27; &quot; &gt; &lt; \`\\"/><meta data-rh=\\"true\\" charSet=\\"utf-8\\"/>"`;

exports[`<meta> tags Server API renders meta tags as string 1`] = `"<meta data-rh=\\"true\\" itemprop=\\"name\\" content=\\"Test name itemprop\\"></meta><meta data-rh=\\"true\\" property=\\"og:type\\" content=\\"article\\"></meta><meta data-rh=\\"true\\" http-equiv=\\"content-type\\" content=\\"text/html\\"></meta><meta data-rh=\\"true\\" name=\\"description\\" content=\\"Test description &amp; encoding of special characters like &#x27; &quot; &gt; &lt; \`\\"></meta><meta data-rh=\\"true\\" charset=\\"utf-8\\"></meta>"`;

exports[`<meta> tags Server Declarative API renders meta tags as React components 1`] = `"<meta data-rh=\\"true\\" itemProp=\\"name\\" content=\\"Test name itemprop\\"/><meta data-rh=\\"true\\" property=\\"og:type\\" content=\\"article\\"/><meta data-rh=\\"true\\" http-equiv=\\"content-type\\" content=\\"text/html\\"/><meta data-rh=\\"true\\" name=\\"description\\" content=\\"Test description &amp; encoding of special characters like &#x27; &quot; &gt; &lt; \`\\"/><meta data-rh=\\"true\\" charSet=\\"utf-8\\"/>"`;

exports[`<meta> tags Server Declarative API renders meta tags as string 1`] = `"<meta data-rh=\\"true\\" itemprop=\\"name\\" content=\\"Test name itemprop\\"></meta><meta data-rh=\\"true\\" property=\\"og:type\\" content=\\"article\\"></meta><meta data-rh=\\"true\\" http-equiv=\\"content-type\\" content=\\"text/html\\"></meta><meta data-rh=\\"true\\" name=\\"description\\" content=\\"Test description &amp; encoding of special characters like &#x27; &quot; &gt; &lt; \`\\"></meta><meta data-rh=\\"true\\" charset=\\"utf-8\\"></meta>"`;
